pyomo/contrib/doe/fim_doe.py

```diff
@@ -0,0 +1,2582 @@
+            for count, para in enumerate(self.param_name):
+                list_jac = []
+                for i in range(len(all_perturb_measure[0])):
+                    if self.scale_nominal_param_value:
+                        sensi = -(all_perturb_measure[count][i] - all_base_measure[count][i]) / self.step * self.scale_constant_value
+                    else:
+                        sensi = -(all_perturb_measure[count][i] - all_base_measure[count][i]) / self.step /self.param_init[para] * self.scale_constant_value
+                    list_jac.append(sensi)
+                # get Jacobian dict, keys are parameter name, values are sensitivity info
+                jac[para] = list_jac
+
+        # check if another prior experiment FIM is provided other than the user-specified one
+        if specified_prior is None:
+            prior_in_use = self.prior_FIM
+        else:
+            prior_in_use = specified_prior
+
+        # Assemble and analyze results
+        FIM_analysis = FIM_result(self.param_name, self.measure, jacobian_info=None, all_jacobian_info=jac,
+                                    prior_FIM=prior_in_use, store_FIM=FIM_store_name, scale_constant_value=self.scale_constant_value)
+
+        time11 = time.time()
+        self.logger.info('Build time with sequential_sipopt or kaug mode [s]:', sum(time_allbuild))
+        self.logger.info('Solve time with sequential_sipopt or kaug mode [s]:', sum(time_allsolve))
+        self.logger.info('Total wall clock time [s]:', time11-time00)
+
+        self.jac = jac
+        FIM_analysis.build_time = sum(time_allbuild)
+        FIM_analysis.solve_time = sum(time_allsolve)
+
+        return FIM_analysis
+
+    def __direct_kaug(self):
+        time00 = time.time()
+        # create scenario class for a base case
+        scena_gen = Scenario_generator(self.param_init, formula=None, step=self.step)
+        scenario_all = scena_gen.simultaneous_scenario()
+
+        # create model
+        time0_build = time.time()
+        mod = self.create_model(scenario_all, args=self.args)
+        time1_build = time.time()
+        time_build = time1_build - time0_build
+
+        # discretize if needed
+        if self.discretize_model is not None:
+            mod = self.discretize_model(mod)
+
+        # get all time
+        t_all = []
+        for t in mod.t:
+            t_all.append(t)
+
+        # add objective function
+        mod.Obj = pyo.Objective(expr=0, sense=pyo.minimize)
+
+        # Check if measurement time points are in this time set
+        # Also correct the measurement time points
+        # For e.g. if a measurement time point is 0.0 in the model but is given as 0, it is corrected here
+        measurement_accurate_time = self.flatten_measure_timeset.copy()
+        for j in self.flatten_measure_name:
+            for no_t, tt in enumerate(self.flatten_measure_timeset[j]):
+                if tt not in t_all:
```

Comment:
It is REALLY slow to test for membership in a list (a linear time operation).  Testing for membership in the original set is constant time.

As currently implemented, you should do `if tt not in mod.t`, but in other comments I am encouraging you to not hard-code `t` as the time set name.  Assuming that the user gave you the time set name as `time_set_name`, then after creating the model, you would want to do `time_set = getattr(mod, time_set_name)`, and then here it would be `if tt not in time_set`

Reply:
changed the code to allow the user to give a `time_set_name` to the time set, not hard coded `t`.